src/transformers/modeling_utils.py

```diff
@@ -3145,7 +3145,7 @@ def from_pretrained(
             config = copy.deepcopy(config)
 
             kwarg_attn_imp = kwargs.pop("attn_implementation", None)
-            if kwarg_attn_imp is not None and config._attn_implementation != kwarg_attn_imp:
+            if kwarg_attn_imp is not None:
```

Comment:
Just from this diff it's not obvious at all to me why this works - when `kwarg_attn_imp` is not None. Previously: 

* Either `config._attn_implementation != kwarg_attn_imp`, in which case we set `config._attn_implementation = kwarg_attn_imp`
* Or `config._attn_implementation == kwarg_attn_imp`, in which case there's no need to set `config._attn_implementation = kwarg_attn_imp` as they're already equal. 

From the tests it works, but it's not clear why. Perhaps @fxmarty can help shed some light

Reply:
@amyeroberts a detailed explanation is here: https://github.com/huggingface/transformers/issues/30298#issuecomment-2061845587

Basically the `_attn_implementation` setter (here https://github.com/huggingface/transformers/blob/e4ea19b958c89d61e42461fac6ac8441787121f8/src/transformers/configuration_utils.py#L406-L420) is never called in this case, resulting in SDPA attention being used although it should be eager.